scripts/migrateContacts.js

```diff
@@ -0,0 +1,168 @@
+
+const { mkAPI } = require('cozy-ach')
+const { getWithInstanceLogger } = require('cozy-ach/libs/utils')
+
+const DOCTYPE_CONTACTS = 'io.cozy.contacts'
+const DOCTYPE_CONTACTS_ACCOUNT = 'io.cozy.contacts.accounts'
+
+function fixCozy(contact) {
+  if (contact.cozy === '') {
+    const fixedContact = {
+      ...contact,
+      cozy: []
+    }
+
+    return {
+      contact: fixedContact,
+      cozyFixed: false,
+      emptyCozyFixed: true
+    }
+  } else if (typeof contact.cozy === 'string') {
+    const fixedContact = {
+      ...contact,
+      cozy: [
+        {
+          url: contact.cozy,
+          primary: true
+        }
+      ]
+    }
+
+    return {
+      contact: fixedContact,
+      cozyFixed: true,
+      emptyCozyFixed: false
+    }
+  }
+
+  return {
+    contact,
+    cozyFixed: false,
+    emptyCozyFixed: false
+  }
+}
+
+function fixEmail(contact) {
+  if (typeof contact.email === 'string') {
+    const fixedContact = {
+      ...contact,
+      email: [
+        {
+          address: contact.email,
+          primary: true
+        }
+      ]
+    }
+
+    return {
+      contact: fixedContact,
+      emailFixed: true
+    }
+  }
+
+  return {
+    contact,
+    emailFixed: false
+  }
+}
+
+function shouldDeleteContact(contact, contactsAccounts) {
+  if (get(contact, 'trashed', false)) {
+    if (get(contact, 'relationships.accounts.data', []).length > 0) {
+      const contactAccountId = contact.relationships.accounts.data[0]._id
+      // get contact account
+      const contactAccount = contactsAccounts.find(
+        account => account._id === contactAccountId
+      )
+
+      return contactAccount.sourceAccount === null
+    }
+  }
+
+  return false
+}
+
+async function doMigrations(dryRun, api, logWithInstance) {
+  const contacts = await api.fetchAll(DOCTYPE_CONTACTS)
+  const contactsAccounts = await api.fetchAll(DOCTYPE_CONTACTS_ACCOUNT)
+  if (contacts.length === 0) {
+    logWithInstance('No contacts, nothing to migrate')
+    return
+  }
+
+  const result = {
+    cozy: 0,
+    cozyEmpty: 0,
+    email: 0
+  }
+
+  const fixedContacts = contacts
+    .map(contact => {
+      const cozyResult = fixCozy(contact)
+      if (cozyResult.cozyFixed) result.cozy++
+      if (cozyResult.emptyCozyFixed) result.cozyEmpty++
+      const { emailFixed, contact: fixedContact } = fixEmail(cozyResult.contact)
+      if (emailFixed) result.email++
+
+      if (cozyResult.cozyFixed || emailFixed) {
+        return fixedContact
+      }
+
+      return undefined
+    })
+    .filter(c => c === undefined)
+
+  const ghosts = contacts.filter(contact =>
+    shouldDeleteContact(contact, contactsAccounts)
+  )
+
+  if (dryRun) {
+    logWithInstance(
+      `Would process ${contacts.length} contacts: \n[cozy]: ${
+        result.cozy
+      } would be fixed.\n[empty cozy]: ${
+        result.cozyEmpty
+      } would be fixed.\n[email]: ${result.email} would be fixed.\n[ghosts]: ${
+        ghosts.length
+      } would be deleted.`
```

Comment:
Not a problem, but the line breaks are weird here? It would be easier to read if they matched the `\n` I think.